### ISSUE TITLE
VirtualPage controller doesn't fall back to methods on copied page controller.

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-cms#1386]
VirtualPage_controller uses the copied page as a failover, when it should use the controller of that page as the failover, which in turn will failover to that copied page.

See https://github.com/silverstripe/silverstripe-subsites/pull/179

Related: Maybe we should get URL actions working as well? (If easy) https://github.com/silverstripe/silverstripe-cms/issues/1101

**Comment by dhensby** [comment]
see https://github.com/silverstripe/silverstripe-cms/pull/1102

was deemed an API change :/

**Comment by dhensby** [comment]
and https://github.com/silverstripe/silverstripe-cms/issues/1099 and https://github.com/silverstripe/silverstripe-cms/pull/905

**Comment by tractorcow** [comment]
Yeah, but #1102 only invokes init on the copied page, not the controller of that page. A lot more is necessary to address this issue.

**Comment by tractorcow** [comment]
This will be an API change... best to do it in master.

**Comment by tractorcow** [comment]
Also see https://github.com/silverstripe/silverstripe-subsites/pull/169

**Comment by tractorcow** [comment]
And https://github.com/silverstripe/silverstripe-subsites/pull/113

**Comment by tractorcow** [comment]
Possibly related: https://github.com/silverstripe/silverstripe-subsites/issues/235 (although to do with template, not methods).

**Comment by tractorcow** [comment]
Possibly related: https://github.com/silverstripe/silverstripe-cms/issues/1336#issuecomment-193352164

**Comment by kinglozzer** [comment]
@tractorcow I was playing around with this a little this afternoon (because of the issue you mentioned about URL actions not working) before I saw your suggested approach 😅. The approach I went for was to simply get rid of `VirtualPage_Controller` and instead use the target page’s controller: setting the VirtualPage instance to be the data record on that.

I.e:
```
MyCustomPage_Controller --> dataRecord = VirtualPage --> CopyContentFrom = MyCustomPage
```

Here’s where I got up to: https://github.com/kinglozzer/silverstripe-cms/commit/a8750065e4ee96e06e42ab92fcc41981b219597e. It would need tidying up, but all the existing VirtualPage unit tests pass and it solves the issue about forwarding `allowed_actions`, `init()` etc. It’d also mean patching subsites, as `SubsiteVirtualPage_Controller extends VirtualPage_Controller`

**Comment by dhensby** [comment]
Sounds like a good approach, actually... Just a shame it's a breaking change...

**Comment by tractorcow** [comment]
>  The approach I went for was to simply get rid of VirtualPage_Controller and instead use the target page’s controller: setting the VirtualPage instance to be the data record on that.

Wow, brilliant. :)

**Comment by tractorcow** [comment]
> Here’s where I got up to: kinglozzer@a875006. It would need tidying up, but all the existing VirtualPage unit tests pass and it solves the issue about forwarding allowed_actions, init() etc. It’d also mean patching subsites, as SubsiteVirtualPage_Controller extends VirtualPage_Controller

Please rebase and PR on master?

**Comment by kinglozzer** [comment]
Now that the namespace PRs have been merged, I finally got around to this: https://github.com/silverstripe/silverstripe-cms/pull/1605

**Comment by tractorcow** [comment]
FIxed with #1605
